Moved R dependencies to conda environment, cleaned up R/python dependencies, and updated base Docker/Travis configuration.

samuelklee opened the pull request:
Closes #5022.
Closes #6307.
Closes #5776.

lbergelson commented:
@samuelklee This is failing with "can't find ggplot2 errors"

samuelklee commented:
Yup, I'm doing some digging.  It fails only for `TEST_DOCKER=true`, any idea why that might be?

lbergelson commented:
It seems to imply that ggplot isn't installed in the docker, which is confusing because it should be.

samuelklee commented:
See latest comments in #4250.  Not sure if using conda to manage R dependencies will affect the size of the image, any idea @jamesemery?

samuelklee commented:
OK, I experimented a bit with removing the R install from the base image and adding the R dependencies to the conda environment in a branch and rebased on that.  A few issues that I've run into or that came up in discussion with @jamesemery and @cmnbroad:

-I moved all tests that depend on R into the `python` test group (which should perhaps be renamed to `conda`).  Note that some of these also fall into the `spark` test group---not sure if there is any special Spark setup done for that group, but we should make sure that they don't fail if they're not run with the conda environment.

-@cmnbroad mentioned that some Picard tools that depend on R may break outside of the conda environment if the user does not have the R dependencies.

-When we install R in the base image, we pull in a lot of basic dependencies (e.g., build-essential, various libraries and compilers, etc.)  So when the R install is removed, it looks like many tests begin failing or hanging, perhaps because they are falling back on Java implementations (e.g., AVX PairHMM tests).  We need to determine the dependencies for these tests and install them separately.  Here is the list of packages that get pulled in by the R install: ```autoconf automake autotools-dev binutils bsdmainutils build-essential
 bzip2-doc cdbs cpp cpp-5 debhelper dh-strip-nondeterminism dh-translations
 dpkg-dev fakeroot g++ g++-5 gcc gcc-5 gettext gettext-base gfortran
 gfortran-5 groff-base ifupdown intltool intltool-debian iproute2
 isc-dhcp-client isc-dhcp-common libalgorithm-diff-perl
 libalgorithm-diff-xs-perl libalgorithm-merge-perl libarchive-zip-perl
 libasan2 libasprintf-dev libasprintf0v5 libatm1 libatomic1
 libauthen-sasl-perl libblas-common libblas-dev libblas3 libbz2-dev
 libc-dev-bin libc6-dev libcc1-0 libcilkrts5 libcroco3 libcurl3
 libdns-export162 libdpkg-perl libencode-locale-perl libfakeroot
 libfile-basedir-perl libfile-desktopentry-perl libfile-fcntllock-perl
 libfile-listing-perl libfile-mimeinfo-perl libfile-stripnondeterminism-perl
 libfont-afm-perl libfontenc1 libgcc-5-dev libgdbm3 libgettextpo-dev
 libgettextpo0 libgfortran-5-dev libgfortran3 libgomp1 libhtml-form-perl
 libhtml-format-perl libhtml-parser-perl libhtml-tagset-perl
 libhtml-tree-perl libhttp-cookies-perl libhttp-daemon-perl libhttp-date-perl
 libhttp-message-perl libhttp-negotiate-perl libio-html-perl
 libio-socket-ssl-perl libipc-system-simple-perl libisc-export160 libisl15
 libitm1 libjpeg-dev libjpeg-turbo8-dev libjpeg8-dev liblapack-dev liblapack3
 liblsan0 liblwp-mediatypes-perl liblwp-protocol-https-perl liblzma-dev
 libmail-sendmail-perl libmailtools-perl libmnl0 libmpc3 libmpfr4 libmpx0
 libncurses5-dev libnet-dbus-perl libnet-http-perl libnet-smtp-ssl-perl
 libnet-ssleay-perl libpaper-utils libpaper1 libpcre16-3 libpcre3-dev
 libpcre32-3 libpcrecpp0v5 libperl5.22 libpipeline1 libpng12-dev libquadmath0
 libreadline-dev libreadline6-dev libsigsegv2 libstdc++-5-dev
 libsys-hostname-long-perl libtcl8.6 libtext-iconv-perl libtie-ixhash-perl
 libtimedate-perl libtinfo-dev libtk8.6 libtsan0 libubsan0 libunistring0
 liburi-perl libwww-perl libwww-robotrules-perl libx11-protocol-perl libxaw7
 libxcb-shape0 libxft2 libxml-parser-perl libxml-twig-perl
 libxml-xpathengine-perl libxmu6 libxmuu1 libxpm4 libxss1 libxtables11 libxv1
 libxxf86dga1 linux-libc-dev m4 make man-db manpages manpages-dev netbase
 patch perl perl-modules-5.22 po-debconf python-pkg-resources python-scour
 python-six r-base-core r-base-dev r-doc-html rename tzdata x11-utils
 x11-xserver-utils xdg-utils zip zlib1g-dev```

-Not sure if moving the R install to the conda environment (which is not in the base image) will increase Travis time, but it doesn't appear to from the limited number of builds that have run so far.  At some point we may want to move conda into the base image.  However, I think that this would require that the base be rebuilt with every python code change, which is not optimal.

samuelklee commented:
@droazen I realize you guys will be preoccupied for the next week or so, but I will leave this PR/branch in engine team's capable hands!

samuelklee commented:
Note also that gCNV WDL tests are hanging because of bullet 3, since theano/PyMC3 is probably falling back on non-compiled implementations as well.

samuelklee commented:
I addressed the ggplot2 dependency in #5040 and moved to commit to address #5022 to that PR.

samuelklee commented:
@droazen can you let me know your plans for this branch?

droazen commented:
I've asked @jamesemery to review. It looks like the branch needs a rebase, though.

samuelklee commented:
Thanks!  Just to be clear, the PR is incomplete.  We need to determine the additional dependencies (which were previously installed along with R) required for AVX, etc.

jamesemery commented on src/test/resources/org/broadinstitute/hellbender/utils/hmm/hmm-unit-test-generate-truth.R:
```diff
@@ -0,0 +1,22 @@
+library(HMM)
```
If you do keep these output files, add a comment about how to generate them to the HMM class incase something is updated in the future.

jamesemery commented on README.md:
````diff
@@ -301,25 +302,7 @@ You can download and run pre-built versions of GATK4 from the following places:
-R 
-source("scripts/docker/gatkbase/install_R_packages.R")
-```
+Certain GATK tools may optionally generate plots using the R installation provided within the conda environment.  If you are uninterested in plotting, R is still required by several of the unit tests.  Plotting is currently untested and should be viewed as a convenience rather than a primary output.
````
Perhaps we should leave the "R" tag on the R tests and and change the build.gradle tests to run the "R" tests under the python package? That is probably more intuitive if in the future we want to reenable support for R.

Then under build.gradle (and dockertest.gradle) I would update the line:
```} else if (TEST_TYPE == "python") {
            includeGroups "python"
``` 
to be 
```} else if (TEST_TYPE == "python") {
            includeGroups "python", "R"
``` 
then further down exclude them by default. This is probably clearer as to what is going on. Then here I would add a comment that the unit tests requiring R are disabled by default and to activate them you must run the python tests.

samuelklee replied:
Added "conda" to the test matrix, which runs both "python" and "R" test groups, and changed back the "R" tag as you suggested.

samuelklee commented on src/test/java/org/broadinstitute/hellbender/utils/hmm/HMMUnitTest.java:
```diff
@@ -69,19 +65,12 @@
-
     /**
-     * This method runs the Test models and sequences against R's HMM package to
+     * This method runs the Test models and sequences against results generated using R's HMM package to
```
The R package was only used to generate results for this test code and is not used in production code, so I'd rather not install it if necessary.  For that matter, even the production code that this tests is not in use, we just kept it around in case anyone needed a pure Java HMM in the future.

I can try to clean up this test class a bit though.

samuelklee commented:
Rebased and tweaked a few package versions, some of which are slightly different than those currently in the base (as is the version of R).  Looks like `build-essential` is all you need in the base image for tests to not fallback on slow implementations, but we might want to double check that native dependencies are correct.

Not really sure how urgent this is, and I have to admit I've lost track of the remaining ways R can break our builds.  The possibility of a bad Travis cache (which I think was a common issue in the past) is now prevented by #6454, right?

samuelklee commented:
For a glorious moment, tests were passing, but it looks like I could not vanquish Xbyak in the end---and now I'm seeing https://github.com/broadinstitute/gatk/issues/6513, as an additional insult...

I have made significant progress in otherwise making sure that the conda environment is consistent, but might need to confer with engine team about next steps.  @droazen @lbergelson @jamesemery @cmnbroad any thoughts are appreciated!  I'll also roll back the changes to the test groups, as James suggested, when I get a chance.

samuelklee commented:
Looks like Xbyak might have been successfully squashed by the update to tensorflow 1.15.0, but it's hard to say given its intermittent nature.  Not sure what to do about the Java 11 funkiness.

droazen commented on scripts/gatkcondaenv.yml.template:
```diff
@@ -1,50 +1,57 @@
+                                    #   and that it is used in tensorflow, theano, and other key dependencies
+- conda-forge::theano=1.0.4         # it is unlikely that new versions of theano will be released
+                                    #   verify that this is using numpy compiled against MKL (e.g., by the presence of -lmkl_rt in theano.config.blas.ldflags)
+- defaults::tensorflow=1.15.0       # update only if absolutely necessary, as this may cause conflicts with other core dependencies
```
Just confirming: is this the Intel-optimized version of tensorflow?

samuelklee replied:
Yes, if you run the check mentioned in the comment returns true, indicating that this version of tensorflow uses Intel MKL-DNN.

samuelklee commented:
Also note https://github.com/broadinstitute/gatk/issues/6020.  Not sure what needs to be done to make sure notebooks on Terra are using this environment.

samuelklee commented on .travis.yml:
```diff
@@ -1,24 +1,21 @@
-language: java
-dist: trusty
+language: generic
```
Note to self: changes here were borrowed from Louis's branch, might have introduced Java 11 issues?

lbergelson replied:
I think we can roll back some of these changes, they were needed because of a travis bug essentially, and they claim to have fixed the bug.  We should be able to specify openjdk8 on bionic now like before.

lbergelson replied:
OracleJDK 8 is just gone from bionic, there's no easy way to install it without a bunch of nonesense.  I think we can just drop the oracle jdk build completely.

samuelklee commented on Dockerfile:
```diff
@@ -43,25 +43,19 @@ WORKDIR /root
-    rm $DOWNLOAD_DIR/miniconda.sh
 WORKDIR /gatk
 ENV PATH $CONDA_PATH/envs/gatk/bin:$CONDA_PATH/bin:$PATH
+ENV PYTHONDONTWRITEBYTECODE=true
```
Note to self: check whether this and/or below affects size, clean up if not.

samuelklee replied:
Not sure if this affects anything at Docker build time, but I'll leave it in.

samuelklee replied:
Hmm, this might slow down the gCNV Travis tests.

samuelklee commented on scripts/gatkcondaenv.yml.template:
```diff
@@ -1,50 +1,57 @@
-- zlib=1.2.11=0
+
+# core python dependencies
+- conda-forge::python=3.6.10        # do not update
```
Note to self: pin build numbers.

samuelklee replied:
Actually, I guess you can only specify build strings, and doing so would not make this platform agnostic.  (Some of these build strings have made their way into the yml before, but we should probably avoid them in the future.)

lbergelson commented:
Bleh

```
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007f513eecd0f2, pid=6936, tid=6963
#
# JRE version: OpenJDK Runtime Environment (11.0.2+9) (build 11.0.2+9)
# Java VM: OpenJDK 64-Bit Server VM (11.0.2+9, mixed mode, tiered, compressed oops, g1 gc, linux-amd64)
# Problematic frame:
# V  [libjvm.so+0x8fd0f2]  jni_GetByteArrayElements+0x72
#
# Core dump will be written. Default location: Core dumps may be processed with "/usr/share/apport/apport %p %s %c %d %P" (or dumping to /home/travis/build/broadinstitute/gatk/core.6936)
#
# An error report file with more information is saved as:
# /home/travis/build/broadinstitute/gatk/hs_err_pid6936.log
#
# If you would like to submit a bug report, please visit:
#   http://bugreport.java.com/bugreport/crash.jsp
```

Then exit 134 which is "something is fucked but we don't know what".  

Looks like a seg fault somewhere with some JNI something....

samuelklee commented:
Yeah, I think that's the same one I noted in #6513 (probably should've noted it here instead).  I'm trying to make the gradle log more verbose (but not too verbose, otherwise Travis complains), but I don't know if that'll help pin it down to a particular failing test or not...

samuelklee commented:
Hmm, running `./gradlew --info ...` yields the following snippet:

```
Gradle suite > Gradle test > org.broadinstitute.hellbender.utils.downsampling.ReservoirDownsamplerUnitTest > testReservoirDownsampler[29](TestDataProvider(ReservoirDownsamplerTest: reservoirSize=10000 totalReads=10000 expectedNumReadsAfterDownsampling=10000 expectedNumDiscardedItems=0)) STANDARD_ERROR
    01:40:10.641 WARN  gatk - Running test: TestDataProvider(ReservoirDownsamplerTest: reservoirSize=10000 totalReads=10000 expectedNumReadsAfterDownsampling=10000 expectedNumDiscardedItems=0)
Finished 130000 tests
Finished 140000 tests

Gradle suite > Gradle test > org.broadinstitute.hellbender.utils.pairhmm.VectorPairHMMUnitTest STANDARD_ERROR
    01:40:14.522 WARN  NativeLibraryLoader - Unable to load libgkl_pairhmm_fpga.so from native/libgkl_pairhmm_fpga.so (/tmp/libgkl_pairhmm_fpga17703278887667828152.so: libgkl_pairhmm_shacc.so: cannot open shared object file: No such file or directory)
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007fe1a5cd00f2, pid=6969, tid=6997
#
# JRE version: OpenJDK Runtime Environment (11.0.2+9) (build 11.0.2+9)
# Java VM: OpenJDK 64-Bit Server VM (11.0.2+9, mixed mode, tiered, compressed oops, g1 gc, linux-amd64)
# Problematic frame:
# V  [libjvm.so+0x8fd0f2]  jni_GetByteArrayElements+0x72
#
# Core dump will be written. Default location: Core dumps may be processed with "/usr/share/apport/apport %p %s %c %d %P" (or dumping to /home/travis/build/broadinstitute/gatk/core.6969)
#
# An error report file with more information is saved as:
# /home/travis/build/broadinstitute/gatk/hs_err_pid6969.log
#
# If you would like to submit a bug report, please visit:
#   http://bugreport.java.com/bugreport/crash.jsp
#
Starting process 'Gradle Test Executor 2'. Working directory: /home/travis/build/broadinstitute/gatk Command: /usr/local/lib/jvm/openjdk11/bin/java -Dgatk.spark.debug -Dorg.gradle.native=false -Dsamjdk.compression_level=2 -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -javaagent:build/tmp/expandedArchives/org.jacoco.agent-0.8.4.jar_982888894296538c98d7324f3ca78d8f/jacocoagent.jar=destfile=build/jacoco/test.exec,append=true,inclnolocationclasses=false,dumponexit=true,output=file,jmx=false @/tmp/gradle-worker-classpath15215090933788021522txt -Xms500M -Xmx3500M -Dfile.encoding=UTF-8 -Duser.country=US -Duser.language=en -Duser.variant -ea worker.org.gradle.process.internal.worker.GradleWorkerMain 'Gradle Test Executor 2'
Successfully started process 'Gradle Test Executor 2'

Gradle Test Executor 2 started executing tests.

> Task :test
Finished 150000 tests
```

samuelklee commented:
I can reproduce an intermittent failure locally.  Here's a snippet of the hs_err_pid*.log:

```
Stack: [0x00007f663c024000,0x00007f663c125000],  sp=0x00007f663c121ee0,  free space=1015k
Native frames: (J=compiled Java code, A=aot compiled Java code, j=interpreted, Vv=VM code, C=native code)
V  [libjvm.so+0x8c4cb0]
C  [libgkl_pairhmm_omp6531264563224332212.so+0x66e92]  JavaData::getData(JNIEnv_*, _jobjectArray*&, _jobjectArray*&)+0x222

Java frames: (J=compiled Java code, j=interpreted, Vv=VM code)
j  com.intel.gkl.pairhmm.IntelPairHmm.computeLikelihoodsNative([Ljava/lang/Object;[Ljava/lang/Object;[D)V+0
j  com.intel.gkl.pairhmm.IntelPairHmm.computeLikelihoods([Lorg/broadinstitute/gatk/nativebindings/pairhmm/ReadDataHolder;[Lorg/broadinstitute/gatk/nativebindings/pairhmm/HaplotypeDataHolder;[D)V+4
j  org.broadinstitute.hellbender.utils.pairhmm.VectorLoglessPairHMM.computeLog10Likelihoods(Lorg/broadinstitute/hellbender/utils/genotyper/LikelihoodMatrix;Ljava/util/List;Ljava/util/Map;)V+221
j  org.broadinstitute.hellbender.utils.pairhmm.VectorPairHMMUnitTest.testLikelihoodsFromHaplotypes(Lorg/broadinstitute/hellbender/utils/pairhmm/PairHMM;Ljava/lang/Boolean;)V+297
v  ~StubRoutines::call_stub
```

I've experimented with changing the testing order of the implementations (e.g., from AVX, OMP, FPGA to OMP, AVX, FPGA), and I've seen failures in both AVX and OMP (i.e., libgkl_pairhmm*.so and libgkl_pairhmm_omp*.so)---but only when the failing implementation is loaded second.  Not sure if this is a real pattern or if it will be useful in debugging, but it seems to be persistent.

samuelklee commented:
@lbergelson @droazen not sure how to proceed here...looks like some GKL + Java 11 funkiness.

Are we officially supporting Java 11 already?

samuelklee commented:
Hmm, actually, could this be a problem due to the way the native libraries are loaded in the test code?  Note that we first cycle through all implementations in the DataProvider, loading the respective library for each implementation via the `synchronized boolean load` method in the `NativeLibraryLoader`.

I'm not really that familiar with concurrency in Java (nor loading native libraries, for that matter), but it seems that the intermittent failure goes away when I refactor the test to remove the DataProvider (by just looping through the implementations in the test method).

Perhaps related to https://github.com/broadinstitute/gatk/issues/5339?

samuelklee commented:
Uncompressed size of the Docker goes up from 4.12GB to 4.76GB.  The correct Tensorflow package + MKL + R contribute most of the extra bulk.

However, using `conda clean -afy` (as opposed to `conda clean -ay`, which we were doing before) and additionally deleting *.a and *.pyc files after the conda install saves ~700MB in the end (without these, the image goes up to 5.43GB).

lbergelson commented:
@samuelklee Would it be possible to remove the .py files instead of the pyc?  I think they're larger.

lbergelson commented:
I don't understand what's going on with the intel segfault at all, but I'm glad you found a solution...

samuelklee commented:
Thanks @lbergelson here's the breakdown:

```
(gatk) root@41051d0bf437:/gatk# find /opt/miniconda/ -follow -type f -name '*.pyc' | xargs du -hs --total | tail -1
9.8M    total
(gatk) root@41051d0bf437:/gatk# find /opt/miniconda/ -follow -type f -name '*.py' | xargs du -hs --total | tail -1
19M     total
(gatk) root@41051d0bf437:/gatk# find /opt/miniconda/ -follow -type f -name '*.a' | xargs du -hs --total | tail -1
276M    total
```

I don't think we actually want to be deleting the *.py files, right?  Were you thinking of something else?

samuelklee commented:
OK, verified that numerical results in the gCNV WDL tests match 4.1.6.0.  Not sure if @lucidtronix needs to run similar checks for the CNN.

Other than needing to push a new base image, I think this branch is ready to go.  However, as we discussed, we should highlight major changes (e.g., the need to run R plotting tools in the conda environment) in the release notes.  @droazen can you assign a final reviewer?

samuelklee commented:
@lbergelson any chance you can take a look at this in the next week or so?

lbergelson commented on scripts/gatkcondaenv.yml.template:
```diff
@@ -1,50 +1,61 @@
+- conda-forge::theano=1.0.4         # it is unlikely that new versions of theano will be released
+                                    #   verify that this is using numpy compiled against MKL (e.g., by the presence of -lmkl_rt in theano.config.blas.ldflags)
+- defaults::tensorflow=1.15.0       # update only if absolutely necessary, as this may cause conflicts with other core dependencies
+                                    #   verify that this is using numpy compiled against MKL (e.g., by checking tensorflow.pywrap_tensorflow.IsMklEnabled())
```
Should we actually run this command during the base image build and fail if it's not working?

samuelklee replied:
I thought about adding some stuff to the package version checks in the unit tests, but it's a good idea to fail early at the image build.  So I should add something to the bash script for building the base image?

I'll have to look at the message output by the numpy MKL check, it's not as simple to parse as the one generated by this tensorflow check.

lbergelson commented on scripts/gatkcondaenv.yml.template:
```diff
@@ -1,50 +1,61 @@
 # $condaEnvDescription
 #
+# Only update this environment if there is a *VERY* good reason to do so!
```
This new documentation of how to update this stuff is a great addition.

droazen commented:
@samuelklee @lbergelson Please have a look at this forum issue -- it may indicate a problem that this branch could solve: https://gatk.broadinstitute.org/hc/en-us/community/posts/360061666671-Broken-conda-env-create-n-gatk-f-gatkcondaenv-yml

samuelklee commented:
Thanks @lbergelson! I agree that it might be worth breaking into more layers—could be worth talking to SV team and seeing what lessons they learned in putting together their hierarchy of images.

Also, note that I pushed the install of miniconda into the base, but I did not push down the setup of the GATK conda environment itself (which takes the bulk of the time during the main-image build, as it requires lots of downloading).  I think I commented elsewhere that a good strategy might be to set up the conda environment with the non-GATK python dependencies in the base, and then pip install the GATK python packages in the main image. This would let us make python code changes without having to rebuild the base, but might require a bit of scripting to create a final yml for non-Docker users.

I also agree that it would be nice to cut down the Travis time, might be worth taking a look at other strategies to do that—could save everyone a lot of time!

samuelklee commented:
OK, added an integration test to check that MKL is enabled.  If we move the conda install of these packages into the base image, then we might need to perform these checks elsewhere, e.g., in the bash script for building the image.

Gotta push the base image and update the main Dockerfile.  I'll merge after the weekend unless there are any more comments.  Again, @droazen please be sure to highlight that R plotting will now require the conda environment in the release notes!

samuelklee commented:
Just noting for posterity that removing build strings in the conda YML seems to have improved portability to different OS environments: https://gatk.broadinstitute.org/hc/en-us/community/posts/360061666671-Broken-conda-env-create-n-gatk-f-gatkcondaenv-yml
